akka-http-core/src/main/scala/akka/http/impl/engine/server/HttpServerBluePrint.scala

```diff
@@ -173,6 +184,106 @@ private[http] object HttpServerBluePrint {
+      setHandler(requestOut, new OutHandler {
+        def onPull(): Unit = pull(requestIn)
+        override def onDownstreamFinish() = cancel(requestIn)
+      })
+      setHandler(responseIn, new InHandler {
+        def onPush(): Unit = {
+          openTimeouts.head.clear()
+          openTimeouts = openTimeouts.tail
+          push(responseOut, grab(responseIn))
+        }
+        override def onUpstreamFinish() = complete(responseOut)
+        override def onUpstreamFailure(ex: Throwable) = fail(responseOut, ex)
+      })
+      setHandler(responseOut, new OutHandler {
+        def onPull(): Unit = pull(responseIn)
+        override def onDownstreamFinish() = cancel(responseIn)
+      })
+    }
+  }
+
+  private class TimeoutSetup(val timeoutBase: Deadline,
+                             val scheduledTask: Cancellable,
+                             val timeout: Duration,
+                             val handler: HttpRequest ⇒ HttpResponse)
+
+  private class TimeoutAccessImpl(request: HttpRequest, initialTimeout: FiniteDuration, requestEnd: Future[Unit],
+                                  trigger: AsyncCallback[(TimeoutAccess, HttpResponse)], materializer: Materializer)
+    extends AtomicReference[Future[TimeoutSetup]] with TimeoutAccess with (HttpRequest ⇒ HttpResponse) { self ⇒
+    import materializer.executionContext
+
+    set {
+      requestEnd.fast.map(_ ⇒ new TimeoutSetup(Deadline.now, schedule(initialTimeout, this), initialTimeout, this))
+    }
+
+    def apply(request: HttpRequest) = HttpResponse(StatusCodes.ServiceUnavailable, entity = "The server was not able " +
+      "to produce a timely response to your request.\r\nPlease try again in a short while!")
+
+    def clear(): Unit = // best effort timeout cancellation
+      get.fast.foreach(setup ⇒ if (setup.scheduledTask ne null) setup.scheduledTask.cancel())
+
+    def updateTimeout(timeout: Duration): Unit = update(timeout, null: HttpRequest ⇒ HttpResponse)
+    def updateHandler(handler: HttpRequest ⇒ HttpResponse): Unit = update(null, handler)
+    def update(timeout: Duration, handler: HttpRequest ⇒ HttpResponse): Unit = {
+      val oldSetup = get
+      val newSetup = Promise[TimeoutSetup]()
+      if (compareAndSet(oldSetup, newSetup.future)) {
+        for (old ← oldSetup.fast) {
+          if ((old.scheduledTask eq null) || old.scheduledTask.cancel()) {
+            val newHandler = if (handler eq null) old.handler else handler
+            val newTimeout = if (timeout eq null) old.timeout else timeout
+            val newScheduling = newTimeout match {
+              case x: FiniteDuration ⇒ schedule(old.timeoutBase + x - Deadline.now, newHandler)
+              case _                 ⇒ null // don't schedule a new timeout
+            }
+            newSetup.success(new TimeoutSetup(old.timeoutBase, newScheduling, newTimeout, newHandler))
+          } // else: too late, the previously set timeout cannot be cancelled anymore
```

Comment:
Just realized that we have the risk for a memory leak here:
If another `update` call comes in after the previous one has failed here it will sit on the never-to-be-fulfilled promise forever.
I'll get that fixed.